### ISSUE TITLE
Confirm keyboard quit

### ISSUE THREAD

**Issue by tavurth** [godotengine/godot-proposals#6915]
### Describe the project you are working on

Day-to-day godot usage

### Describe the problem or limitation you are having in your project

I often use CMD+Q to quit the running app.

Quite often the window will appear and then disappear as the game crashes because of some code bug.

In this case if I hit CMD+Q the whole godot editor exits.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Please add a confirm to quit

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

When I try to quit the editor using a shortcut, it will ask me if I'm sure I want to quit

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No

### Is there a reason why this should be core and not an add-on in the asset library?

It's Godot

**Comment by Zireael07** [comment]
This seems a neverending wheel. Some people requested confirm quit, so it was added, then people said modern apps do not have this, so this was removed, and now people ask for confirm quit again....

**Comment by tavurth** [comment]
I wonder why it's not just left as an option in the editor preferences if that's the case 🤔

**Comment by Calinou** [comment]
- Duplicate of https://github.com/godotengine/godot-proposals/issues/6190.

See https://github.com/godotengine/godot/pull/32321 as for why there is no setting anymore.

**Comment by tavurth** [comment]
Hmm perhaps then it's because I primarily edit code from a separate IDE (emacs).

In that case perhaps this is rather a bug where the editor is not detecting externally changed files, and so exits without showing a confirm dialog?

In any-case, having an escape hatch for this (always confirm quit) would be a nice toggle to have in the editor preferences.

**Comment by tavurth** [comment]
I really like @akien-mga's suggestion from one of those PRs you linked:

A better approach would be to have a quit confirmation dialog enabled by default, but with a checkbox in the dialog to disable it ("Don't ask me again" style).

**Comment by Calinou** [comment]
> Am I correct in understanding that since https://github.com/godotengine/godot/pull/32321 there is no way to revert this behaviour from the preferences?

Indeed, there isn't.

> see there is no `Quit` shortcut since it uses the OS's quit.

We're following macOS's human interface guidelines here, which means using the built-in shortcut instead of providing our own (like we do on Windows and Linux).

**Comment by tavurth** [comment]
> If you save the script in an external editor, there are no unsaved changes within the Godot editor. If you don't save the script, there is no way for Godot tocknow you didn't save the script yet (and you won't lose data if you only close the Godot editor).

Aha so it's finally a bug 🤔 
I guess a regression also as in 3.5 this did not occur

**Comment by Zireael07** [comment]
> If we make all crashes spawn a visible dialog, then your Cmd + Q key press will go to the alert dialog, not the editor window. This will prevent the editor from being accidentally exited.

Per my question in the PR, there will not be an alert dialog in the Android port. So it can still be accidentally exited

**Comment by tavurth** [comment]
@Calinou this feature would be pretty useful for me.

Since I don't see any escape hatch for this behaviour, what would you think if I were to revert this PR but set toggle to false by default?

https://github.com/godotengine/godot/pull/32321/files#diff-985565fb1f983acb2d87dbe8931c929d0f46fdecb1d99a31e495b151c0f52186L360

**Comment by Calinou** [comment]
> @Calinou this feature would be pretty useful for me.
> 
> Since I don't see any escape hatch for this behaviour, what would you think if I were to revert this PR but set toggle to false by default?
> 
> [godotengine/godot#32321 (files)](https://github.com/godotengine/godot/pull/32321/files#diff-985565fb1f983acb2d87dbe8931c929d0f46fdecb1d99a31e495b151c0f52186L360)

This is not a real fix, but a band-aid solution. I suggest you apply this change locally, but don't open a PR.

**Comment by tavurth** [comment]
> This is not a real fix, but a band-aid solution.

Then better to have the "don't ask this again" option? What would be a good fix here?

**Comment by Calinou** [comment]
> What would be a good fix here?

What I consider an ideal fix in terms of usability is the one outlined here: https://github.com/godotengine/godot-proposals/issues/6915#issuecomment-1555883071

It does not nag you excessively (not even once), but it prevents situations where you accidentally quit the editor.

**Comment by tavurth** [comment]
> As for script errors (which are not crashes), we could add a quit confirmation when a project is running (or paused) from the editor. This makes sense as you can't reattach to the running project after quitting the editor.

Ok makes sense, sorry I missed it before.

So just reverting this PR and change this line to check and see if there is a running instance?
https://github.com/godotengine/godot/pull/32321/files#diff-b88f0bde58b731e8dd353707c8011fb7a1e452f0db2c9f15f4b7905d60e5e52eR2600

**Comment by Calinou** [comment]
> So just reverting this PR and change this line to check and see if there is a running instance?
> [godotengine/godot#32321 (files)](https://github.com/godotengine/godot/pull/32321/files#diff-b88f0bde58b731e8dd353707c8011fb7a1e452f0db2c9f15f4b7905d60e5e52eR2600)

Yes, although the message needs to be modified accordingly to denote that a project is currently running from the editor.

**Comment by hhhapz** [comment]
An argument I've seen in this issue and the issue in which originally removed this feature goes to the tone of:

> It is unreasonable to hide such an option in the settings since no one will know of its existence if it is disabled by default.

We can clearly see that from the number of people who don't like this behaviour enough to create an issue (and many more who have not) that this is behaviour users find frustrating.

However, 'Editor Settings' pane has a handy search feature! I don't mind if this is disabled by default. Before I came to this issue, the first step I took was to search for 'quit' in the editor settings pane, and I'm certain other users would do this too if they needed this functionality. User's wont need look through "hundreds of editor settings" to discover this. Other software like [Jetbrains' Editor Suite](https://intellij-support.jetbrains.com/hc/en-us/community/posts/115000376504-How-to-undo-Do-not-ask-again-when-exiting) and [VSCode](https://github.com/microsoft/vscode/issues/14710) also provide this functionality.

I use Linux with a tiling window manager, and keyboard focus follows to the application my cursor is over. Sometimes I can be sloppy and I don't properly have my mouse over my game and close the editor instead.

It seems that the current agreement this issue has come to is to address accidental closes if the game is running within the editor. Is there an open PR for this? Otherwise I can contribute.

However, I would also request reconsideration to add a setting to always prompt.

**Comment by Calinou** [comment]
> It seems that the current agreement this issue has come to is to address accidental closes if the game is running within the editor. Is there an open PR for this? Otherwise I can contribute.

Not currently, but feel free to open a pull request for this :slightly_smiling_face:
